src/integrations/iron_handlers.rs

```diff
@@ -77,59 +85,77 @@ impl<'a, CtxFactory, Query, Mutation, CtxT>
+        itry!(req.body.read_to_string(&mut request_payload));
+        let json_data =
+            match serde_json::from_str::<Json>(&*request_payload) {
+                Ok(json) => json,
+                Err(err) => {
+                    let error = IronError::new(
+                        Box::new(GraphQlIronError::Serde(err)),
+                        (status::BadRequest, "No JSON object was decoded."));
+                    return Err(error)
+                }
+            };
+        match json_data {
+            Json::Object(json_obj) => {
+                let mut query = None;
+                let mut variables = Variables::new();
+
+                for (k, v) in json_obj {
+                    if k == "query" {
+                        query = serde_json::to_string(&v)
+                            .ok()
+                            .map(|query| query.replace("\"", ""));
+                    }
+                    else if k == "variables" {
+                        variables = InputValue::from_json(v).to_object_value()
+                            .map(|o| o.into_iter().map(|(k, v)| (k.to_owned(), v.clone())).collect())
+                            .unwrap_or_default();
+                    }
+                }
+                let query = iexpect!(query);
+                self.execute(req, &query, &variables)
             }
-            else if k == "variables" {
-                variables = InputValue::from_json(v).to_object_value()
-                    .map(|o| o.into_iter().map(|(k, v)| (k.to_owned(), v.clone())).collect())
-                    .unwrap_or_default();
+            _ => {
+                let error = IronError::new(
+                    Box::new(GraphQlIronError::IO(IoError::new(ErrorKind::InvalidData,
+                                                                "Was able parse a JSON item but it\
+                                                                 was not an object as expected."))),
+                            (status::BadRequest, "No JSON object was decoded."));
+                Err(error)
             }
         }
-
-        let query = iexpect!(query);
-
-        self.execute(req, &query, &variables)
     }
 
     fn execute(&self, req: &mut Request, query: &str, variables: &Variables) -> IronResult<Response> {
         let context = (self.context_factory)(req);
-        let result = execute(query, None, &self.root_node, variables, &context);
-
+        let result = execute_query(query, None, &self.root_node, variables, &context);
         let content_type = "application/json".parse::<Mime>().unwrap();
-        let mut map = BTreeMap::new();
+        let mut map = HashMap::new();
 
         match result {
             Ok((result, errors)) => {
-                map.insert("data".to_owned(), result.to_json());
+                let response_data = serde_json::to_value(result)
+                    .expect("Failed to convert response data to JSON.");
```

Comment:
Wasn't sure how you, @mhallin, wanted handed handle errors at this points especially since the response has probably been parsed already and should only fail if the developer messes up.